<launch>	

<!-- Launch rviz -->
<node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz" args="-d $(find pointcloud_registration)/cetran_velodyne.rviz" />

<!-- LAUNCH RSLIDAR DRIVER HERE 
	<node  name="rslidar_node" pkg="rslidar_driver" type="rslidar_node" output="screen" >
		<param name="model" value="RS32"/>
		<param name="device_ip" value="192.168.200.204"/>
		<param name="port" value="3207"/>
	</node>

	<node  name="cloud_node" pkg="rslidar_pointcloud" type="cloud_node" output="screen" >
		<param name="model" value="RS32"/>
		<param name="curves_path" value="$(find rslidar_pointcloud)/ConfigrationFiles/curves.csv" />
		<param name="angle_path" value="$(find rslidar_pointcloud)/ConfigrationFiles/angle.csv" />
		<param name="channel_path" value="$(find rslidar_pointcloud)/ConfigrationFiles/ChannelNum.csv" />
		<param name="curves_rate_path" value="$(find rslidar_pointcloud)/ConfigrationFiles/CurveRate.csv" />

	</node>
-->
<!--	IF USING 3D LOCALIZATION WITH 3D LIDAR

	<node pkg="pointcloud_registration" type="icp_localize" name="icp_localize" output="screen">
		<param name="map_location" value="$(arg map_location)" />
		<param name="base_frame_id" value="iMiev/base_link" />
		<param name="odom_frame_id" value="iMiev/odom" />
		<param name="global_frame_id" value="iMiev/map" />
		<param name="icp_iteration" value="5" />
		<param name="icp_distance" value="5.0" />
		<param name="icp_filter_size" value="0.5" />
		<remap from="input" to="/rslidar_points" />
	</node>

	<node pkg="pointcloud_registration" type="show_map" name="show_map" output="screen">
		<param name="map_location" value="$(arg map_location)" />
		<param name="global_frame_id" value="iMiev/map" />
	</node>
-->

<!--    IF USING 2D LOCALIZATION WITH 3D LIDAR -->
	
	<node pkg = "pointcloud_filter" type="pointcloud_filter_vertical_feature" name="filter_vertical_feature">
		<remap from = "cloud_input" to = "/rslidar_points" />
		<remap from = "vertical_points" to = "/iMiev/pcl_pointnormal_out" />
		<param name = "base_frame" value = "iMiev/base_link" />
	</node>

<!---->
	<node pkg="amcl_pc" type="amcl_pc" name="amcl_pc" output="screen">
		<remap from="pc" to="/iMiev/pcl_pointnormal_out"/>
		<param name="odom_frame_id" value="iMiev/odom" />
		<param name="base_frame_id" value="iMiev/base_link" />
		<param name="global_frame_id" value="iMiev/map" />
		<param name="laser_model_type" value="pc_likelihood_field" />
	</node>


<!--    IF USING 2D LOCALIZATION WITH 2D LIDAR 

	<node pkg="amcl_pc" type="amcl" name="amcl_pc" output="screen">
		<remap from="scan" to="/iMiev/front_top_scan"/>
		<remap from="pc" to="/iMiev/pcl_pointnormal_out"/>
		<param name="odom_frame_id" value="iMiev/odom" />
		<param name="base_frame_id" value="iMiev/base_link" />
		<param name="global_frame_id" value="iMiev/map" />
		<param name="laser_model_type" value="pc_likelihood_field" />
	</node>

	<node pkg="norm_virtual_sensor" type = "norm_virtual_sensor" name="norm_virtual_sensor" output = "screen">
		<remap from = "scan_in" to="/iMiev/front_top_scan" />
		<remap from = "pcl_pointnormal_out" to = "/iMiev/pcl_pointnormal_out" />
		<param name = "odom_frame_id" value="iMiev/odom" />
		<param name = "baselink_frame_id" value="iMiev/base_link" />
		<param name = "odom_baselink_frame_id" value="iMiev/odom_baselink" />
	</node>

	<node pkg="norm_virtual_sensor" type="odom_baselink" name="odom_baselink" output = "screen">
		<param name = "odom_frame_id" value="iMiev/odom" />
		<param name = "baselink_frame_id" value = "iMiev/odom_baselink" />
		<remap from = "odom" to = "iMiev/odom" />
	</node>
-->
<!---->	<node pkg="map_server" type="map_server" name="grid_cell_map" args="$(find pointcloud_registration)/map/newmap.yaml">
		<param name="frame_id" value="iMiev/map" />
	</node> 

	<node name="global_scan" pkg = "pointcloud_registration" type="scan_matching" output="screen">
		<remap from="input" to="/vertical_points" />
		<remap from="full_map" to="cloud_map"/>
		<param name="icp_iteration" value="100" />
		<param name="icp_distance" value="10.0" />
		<param name="icp_filter_size" value="0.2" />
		<param name="global_frame_id" value="iMiev/map" />
		<param name="base_frame_id" value="iMiev/base_link" />
	</node>

	<node  name="mask_filter" pkg="pointcloud_filter" type="pointcloud_filter_mask" output="screen">
		<remap from="floor_points" to="/cloud_map"/>
		<remap from="map_mask" to="/iMiev/map" />
		<param name="map_frame" value="iMiev/map" />
		<param name="filter_integer" value="100" />
	</node>

	<node name="vertical_filter_for_global" pkg= "pointcloud_filter" type="pointcloud_filter_vertical_feature" output="screen">
		<remap from="cloud_input" to="/rslidar_points" />
		<param name="filter_radius" value="100.0"/>
		<param name="repeat" value="false"/>
		<param name="base_frame" value="iMiev/base_link" />
	</node>  




	
</launch>
